adapters/repos/db/vector/hnsw/search.go

```diff
@@ -557,22 +557,33 @@ func (h *hnsw) knnSearchByVector(searchVec []float32, k int,
 	}
 
 	if h.shouldRescore() {
-		ids := make([]uint64, res.Len())
-		i := len(ids) - 1
-		for res.Len() > 0 {
-			res := res.Pop()
-			ids[i] = res.ID
-			i--
+		ids := make([]uint64, 0, res.Len())
+		for _, item := range res.Items() {
+			ids = append(ids, item.ID)
 		}
 		res.Reset()
 		for _, id := range ids {
 			dist, _, _ := h.distanceFromBytesToFloatNode(byteDistancer, id)
-			res.Insert(id, dist)
-			if res.Len() > ef {
-				res.Pop()
+			if res.Len() == k {
+				if dist < res.Top().Dist {
```

Comment:
Do we not want to include a result that is == `res.Top().Dist`?

Reply:
Depends on how we break ties. First in wins, then the incoming new vector only gets in if better, not equal. But I would expect ties to be very rare honestly, unless vectors are equal.